lib/scanner.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 'use strict';
+
+var fs = require('fs');
+var path = require('path');
+var crypto = require('crypto');
+var findit = require('findit');
+
+module.exports = {
+  scan: scan,
+  findJSFiles: findJSFiles,
+  computeHash: computeHash
+};
+
+function scan(baseDir, callback) {
+  findJSFiles(baseDir, function(err, fileList) {
+    if (err) {
+      callback(err);
+      return;
+    }
+    computeHash(fileList, function(err, hash) {
+      callback(err, hash, fileList);
+    });
+  });
+}
+
+/**
+ * This function accept an array of filenames and computes a unique hash-code
+ * based on the contents.
+ *
+ * @param {!Array<string>} fileList array of filenames
+ * @param {!function(?Error, string)} callback error-back style callback
+ *    returning the hash-code.
+ */
+function computeHash(fileList, callback) {
+  var pending = fileList.length;
+  // return a valid, if fake, result when there are no js files to hash.
+  if (pending === 0) {
+    callback(null, 'EMPTY-no-js-files');
+    return;
+  }
+
+  var hashes = [];
+  fileList.forEach(function(filename) {
```

Comment:
The `forEach` callback can take an index as its second argument. Using this will remove the need for pending.